src/azure-cli-core/azure/cli/core/auth/util.py

```diff
@@ -69,10 +79,14 @@ def _generate_login_command(tenant=None, scopes=None, claims_challenge=None):
 
     # Rejected by CAE
     if claims_challenge:
-        # Explicit logout is needed: https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/335
-        return 'az logout\n' + ' '.join(login_command)
+        from azure.cli.core.util import b64encode
+        # Base64 encode the claims_challenge to avoid shell interpretation
+        claims_challenge_encoded = b64encode(claims_challenge)
+        login_command.extend(['--claims-challenge', f'"{claims_challenge_encoded}"'])
 
-    return ' '.join(login_command)
+    # Explicit logout is preferred, making sure MSAL cache is purged:
+    # https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/335
+    return 'az logout\n' + ' '.join(login_command)
```

Comment:
We always ask the user to explicitly log out, in case some commands (such as `az ad`) don't support handling `401` challenge. See https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/335.